### ISSUE TITLE
React Routing

### ISSUE THREAD

**Issue by marco-dasilva** [kyle-mccarthy/nest-next#38]
I usually like to do my research and spend a lot of time doing so before reaching out for help but I'm at a dead end. I must be having tunnel vision at this point but either way here goes.

I understand the reason for handling routes in Nest and routing them to Next but I'm actually looking for Next to handle anything not defined in Nest. I don't know if this is the correct use case but here's an example none the less.

In pages/, I create a file called index.tsx and in nest I add to the AppController a Get('') and Render('Index'). This loads on localhost:3000. In this index.tsx file, I create a link to redirect the user to a new page called Profile (<Link>). Clicking on this link for Profile, it successfully routes the user however when I refresh, Nest kicks in and throws an error.

Is there a solution for this?

In conclusion, if I go to `/asadd`, a non existent route in Nest, I would like to see the Next 404 error page and not the Nest stack trace page.

**Comment by kyle-mccarthy** [comment]
So I haven't actually tried this before, but messed around and think I found a solution. You will want to create your own global ExceptionFilter that passes the request and response to Next rather than handling it in Nest.

I was able to get it working with the following code. Some things to keep in mind is that this will pass on all of your NotFoundExceptions to Next, depending on your use case this should probably be fine. Also it will route based on the filesystem path, if you have a some route /test it resolves to `pages/test.tsx`.

Let me know if this works for you!

```typescript
import {
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  HttpException,
  Inject,
  NotFoundException,
} from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { RenderModule, RenderService, RequestHandler } from 'nest-next';
import Next from 'next';
import 'reflect-metadata';
import { AppModule } from './application.module';

// Additional global filter to pass 404 errors onto next rather than
// handling in nest.
@Catch(NotFoundException)
export class NextPageFilter implements ExceptionFilter {
  private requestHandler?: RequestHandler;

  constructor(@Inject() private readonly renderService: RenderService) {
    this.requestHandler = this.renderService.getRequestHandler();
  }

  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const res = ctx.getResponse();
    const req = ctx.getRequest();

    if (this.requestHandler) {
      return this.requestHandler(req, res, req.query);
    }

    // the requestHandler technically should always be defined if the nest-next
    // was correctly setup - rethrow just in case
    throw exception;
  }
}

async function bootstrap() {
  const dev = process.env.NODE_ENV !== 'production';
  const app = Next({ dev });

  await app.prepare();

  const server = await NestFactory.create(AppModule);

  const renderer = server.get(RenderModule);
  renderer.register(server, app);

  const service = server.get(RenderService);
  server.useGlobalFilters(new NextPageFilter(service));

  await server.listen(3000);
}

bootstrap();
```

**Comment by marco-dasilva** [comment]
I see it hitting the constructor but it isn't hitting the catch exception - I had a similar implementation of this earlier from a stackoverflow thread so I'll see what I missed there. I am confident I setup nest-next correctly with fastify.

Here's an example screenshot of what I'm seeing without defining the render in Nest. I used your basic example from the repo.
![cannot-get-object](https://user-images.githubusercontent.com/38259976/73550516-bf6e5300-4412-11ea-8984-55b6743e33c7.png)

I am not sure why I didn't think of this but checking the source code of the served page `Cannot GET /test` shows next delivering it with a stack trace but for some reason when I setup the following code in Nest and then attempt to hit `/test` I get the actual 404 that I wanted:
```
@Render('test')
@Get('/test')
````
![404](https://user-images.githubusercontent.com/38259976/73550517-bf6e5300-4412-11ea-9146-8a6575456523.png)

**Comment by kyle-mccarthy** [comment]
Could you try updating the code for that filter to call the request handler like `return this.requestHandler(req, res);`. I noticed a problem with the query not being what next was expecting.

Also I just wanted to confirm two things:

1. Make sure that you attach the global filter to the server after registering the RenderModule

```
const renderer = server.get(RenderModule);
renderer.register(server, app);

const service = server.get(RenderService);
server.useGlobalFilters(new NextPageFilter(service));
```

2. The test file's file location is `/pages/test.tsx`, it shouldn't be in the views folder. Also the naming is case sensitive so it should be lowercase.

**Comment by marco-dasilva** [comment]
Good point! For (1) above, the order mattered - I was incorrectly setting it before it registered. Once it hit the filter, now seeing:
`The "url" argument must be of type string`

**Comment by marco-dasilva** [comment]
I can confirm it's when using fastify.

**Comment by marco-dasilva** [comment]
Solved the issue by returning this on NextPageFilter:
```
return this.requestHandler(res.request.raw, res.res);
```

**Comment by kyle-mccarthy** [comment]
ah yeah, I forgot that fastify modifies the req and res. I believe that you can also do

```
return this.requestHandler(req.raw, res.res)
````

That is how we handled it in render.filter.ts FWIW.

**Comment by jgoux** [comment]
This is exactly my use-case, I was actually surprise that it isn't the default.

Do you think this could be included and enabled via an option?

**Comment by kyle-mccarthy** [comment]
@jgoux I think that should be possible. Just to clarify, you want all the requests that Nest thinks are 404s to be handled by Next's request handler? The only problem that I could see with this, is even if you throw a NotFoundException it would still be forwarded to nextjs. If we were to go this route I think it would resolve #47 and #48.

Also I believe that calling Next tries to start the server, so the exception being thrown is probably because it sees something using the port. Can I ask why you're calling it from NextPageFilter?

**Comment by jgoux** [comment]
@kyle-mccarthy Sorry I meant I'm callling it from both `bootstrap` and the `RenderModule.forRootAsync` in my AppModule.

So it seems to start the Next.js compilation twice in parallel.

If we use the NextPageFilter approach, should we do anything special to the AppModule?

**Comment by kyle-mccarthy** [comment]
@jgoux  Oh yeah so this issue is a little older and uses the deprecated bootstrapping method - you don't need to call it in your bootstrap file anymore. See https://github.com/kyle-mccarthy/nest-next/blob/master/examples/basic/src/main.ts

Also to follow up on my question, are you wanting to have all 404s be handled by next instead of nest handling it? This is a feature I would definitely be open to adding, and it could be controlled through an option like you suggested.

**Comment by jgoux** [comment]
Yes I don't mind having next handling all the 404s. What could be the downside?

As a total outsider I really thought this was the default behaviour, and that you could add specific Nest handlers as you go.

Just to be sure, following the modern bootstraping approach, is this the correct way to register the NextPageFilter? Or can it be moved to AppModule as well?

```typescript
async function bootstrap() {
  const server = await NestFactory.create<NestFastifyApplication>(
    AppModule,
    new FastifyAdapter()
  );

  const service = server.get(RenderService);
  server.useGlobalFilters(new NextPageFilter(service));

  await server.listen(3000);
}
```

**Comment by kyle-mccarthy** [comment]
That is correct, since you need to provide the RenderService for your filter.

And I don't really see any issue with passing the request to next when nest 404s, I will look into providing that option.

**Comment by jgoux** [comment]
This is awesome, thanks **a lot** for your answers and your work on this library! The NestJS ecosystem really seems great. 😄

**Comment by dayrim** [comment]
Any chance you can provide an example project for this ?

**Comment by dayrim** [comment]
Any chance you can provide an example project for this ?

**Comment by kyle-mccarthy** [comment]
@jgoux @dayrim I've made the updates that will allow for you passthrough requests that 404 to next. It's currently on the [next branch](https://github.com/kyle-mccarthy/nest-next/tree/next) and you can preview it by doing a `yarn add nest-next@beta`. There is also an example available here https://github.com/kyle-mccarthy/nest-next/tree/next/examples/passthrough

I've still got to update the main docs, but there are only two changes required to get started.

1. Update the dir property in next's config `dir: resolve(__dirname, '..')`
2. Enable the `passthrough404` option in the `nest-next` config

Both of those are highlighted in the [example's ApplicationModule](https://github.com/kyle-mccarthy/nest-next/blob/next/examples/passthrough/src/application.module.ts).

As a side note, I couldn't think of a great name for the option so I took a pretty literal approach. LMK if you think there is a name that better describes the feature.

**Comment by jgoux** [comment]
Thanks for the change, I'll test it soon!

Is there any way to support `src/pages` as the path? Next.js supports both `pages` and `src/pages` as valid paths, and I prefer to group everything in `src`.

For the naming, maybe `nextFallback`? I don't know, naming is the hardest part. 😄

**Comment by kyle-mccarthy** [comment]
@jgoux I actually think that should be possible by omitting step 1 from my previous comment. The only reason I actually did that is so that next knows it should look in the `src`'s parent directory for the `pages` dir.

**Comment by kyle-mccarthy** [comment]
@jgoux Did you have an opportunity to try this out? If it works for your use case, I can merge this and deploy a new version.

**Comment by jgoux** [comment]
@kyle-mccarthy No sorry I was on another subject. I'll try it this week and let you know. Thanks again for your patience.

**Comment by twigs67** [comment]
@kyle-mccarthy I'm getting the following error:
```
Argument of type '{ passthrough404: boolean; viewsDir: null; }' is not assignable to parameter of type 'Partial<RendererConfig>'.
  Object literal may only specify known properties, and 'passthrough404' does not exist in type 'Partial<RendererConfig>'.
```
Any idea how I can get this working?

Thanks

**Comment by kyle-mccarthy** [comment]
@twigs67 hey thanks for checking this out - I just want to make sure that you have the right version installed since this currently isn't mastered. Did you add `nest-next` with `yarn add nest-next@beta`?

**Comment by twigs67** [comment]
@kyle-mccarthy I did not. Sorry about that - should have read a bit more. Thanks!

**Comment by kyle-mccarthy** [comment]
@twigs67 no problem at all - I actually do want to master this ASAP but have been awaiting feedback. If it works for your use case and works how you'd expect let me know and I'll master it + deploy a new version.

**Comment by twigs67** [comment]
@kyle-mccarthy yes, it did work for me. However, I had to remove `viewsDir` as it was giving me a 404.

**Comment by ralls** [comment]
@kyle-mccarthy do you plan on releasing `9.3.0-beta.0`? It's the only way I am able to implement Auth0.

**Comment by kyle-mccarthy** [comment]
@ralls done! just deployed v10 to npm

**Comment by ralls** [comment]
@kyle-mccarthy Thank you! 👍

**Comment by joshkay** [comment]
Is it possible to filter which 404 reponses are passed through to next?  I would like to prevent my 404 api responses from rendering an error page.
